File: BibliotecaPessoal.BackEnd/Entity/Book.cs
namespace BibliotecaPessoal.BackEnd.Entity;

public class Book
{
    public required string Isbn { get; set; }

    public required string Title { get; set; }

    public string? Author { get; set; }

    public string? Synopsis { get; set; }

    public string? ImageUrl { get; set; }

    public DateOnly RegisteredAt { get; set; }

    public int? PagesTotal { get; set; }

    public int PagesRead { get; set; }

    public decimal? Rating { get; set; }

    public string? Review { get; set; }

    // Derivado, nunca persistido: guardar o progresso criaria uma segunda fonte
    // de verdade que pode discordar de PagesRead/PagesTotal.
    // O padrao "is > 0" cobre null e zero de uma vez, e o cast evita divisao inteira.
    public double Progress => PagesTotal is > 0
        ? Math.Round((double)PagesRead / PagesTotal.Value * 100, 2)
        : 0;
}

File: BibliotecaPessoal.FrontEnd/src/app/core/book.ts
export interface Book {
  isbn: string;
  title: string;
  author: string | null;
  image: string | null;
  progress: number;
  pages_read: number;
  pages_total: number | null;
  rating: number | null;
  review: string | null;
  synopsis: string | null;
}

export interface BookList {
  count: number;
  page_count: number;
  books: Book[];
}

export type SortBy =
  | 'title'
  | 'isbn'
  | 'author'
  | 'rating'
  | 'pages_read'
  | 'pages_total'
  | 'registered_at';

export type FilterBy = 'title' | 'author' | 'isbn';

export interface BookQuery {
  page?: number;
  page_size?: number;
  sort_by?: SortBy;
  sort_direction?: 'asc' | 'desc';
  filter_by?: FilterBy;
  filter?: string;
}
